File: lessons/05-data-structures/data_structures.pl
#!/usr/bin/env swipl
% Lesson 5: Data Structures in Prolog
% Demonstrates lists, terms, and logical structures

:- initialization(main, main).

main :-
    writeln('=== Prolog Data Structures ==='),
    nl,

    % 1. Lists (immutable)
    writeln('1. Lists (Immutable):'),
    Numbers = [1, 2, 3, 4, 5],
    format('  List: ~w~n', [Numbers]),

    % Pattern matching with lists
    [Head|Tail] = Numbers,
    format('  Head: ~w, Tail: ~w~n', [Head, Tail]),

    % List operations
    length(Numbers, Len),
    format('  Length: ~w~n', [Len]),

    reverse(Numbers, Reversed),
    format('  Reversed: ~w~n', [Reversed]),

    % 2. List construction
    nl,
    writeln('2. List Construction:'),
    NewList = [0|Numbers],  % Prepend
    format('  Prepend 0: ~w~n', [NewList]),

    append(Numbers, [6], WithSix),
    format('  Append 6: ~w~n', [WithSix]),

    append([1, 2], [3, 4], Combined),
    format('  Append [1,2] and [3,4]: ~w~n', [Combined]),

    % 3. List predicates
    nl,
    writeln('3. List Predicates:'),
    (member(3, Numbers) ->
        writeln('  member(3, list): true')
    ;
        writeln('  member(3, list): false')
    ),

    nth0(0, Numbers, First),
    format('  nth0(0): ~w~n', [First]),

    % 4. Recursive list operations
    nl,
    writeln('4. Recursive List Operations:'),
    sum_list(Numbers, Sum),
    format('  sum_list: ~w~n', [Sum]),

    % Custom map
    map_double(Numbers, Doubled),
    format('  map_double: ~w~n', [Doubled]),

    % Custom filter
    filter_evens([1, 2, 3, 4, 5, 6], Evens),
    format('  filter_evens: ~w~n', [Evens]),

    % 5. Terms and structures
    nl,
    writeln('5. Terms and Structures:'),
    Person = person('Alice', 30, 'NYC'),
    format('  Person term: ~w~n', [Person]),

    % Pattern matching on structures
    person(Name, Age, City) = Person,
    format('  Destructured: ~w, ~w, ~w~n', [Name, Age, City]),

    % 6. Key-value pairs (association lists)
    nl,
    writeln('6. Association Lists (Key-Value):'),
    Dict = [name-'Bob', age-25, city-'SF'],
    format('  Dict: ~w~n', [Dict]),

    % Lookup
    (member(name-N, Dict) ->
        format('  Lookup name: ~w~n', [N])
    ;
        writeln('  Name not found')
    ),

    % 7. Nested lists
    nl,
    writeln('7. Nested Lists:'),
    Matrix = [[1, 2, 3], [4, 5, 6], [7, 8, 9]],
    format('  Matrix: ~w~n', [Matrix]),

    nth0(1, Matrix, Row),
    nth0(2, Row, Element),
    format('  matrix[1][2]: ~w~n', [Element]),

    % 8. Difference lists (advanced)
    nl,
    writeln('8. Difference Lists (Efficient Append):'),
    writeln('  Difference lists enable O(1) append'),
    DiffList1 = [1, 2, 3|Tail]-Tail,
    format('  Diff list: ~w~n', [DiffList1]),

    % 9. Unification (Prolog's power)
    nl,
    writeln('9. Unification (Pattern Matching):'),
    Point = point(X, Y),
    Point = point(3, 4),
    format('  Unified point(X,Y) with point(3,4): X=~w, Y=~w~n', [X, Y]),

    % 10. Key insights
    nl,
    writeln('10. Key Insights:'),
    writeln('  - Lists are fundamental to Prolog'),
    writeln('  - Pattern matching via unification'),
    writeln('  - Lists are immutable'),
    writeln('  - Terms provide structured data'),
    writeln('  - Recursive processing is natural'),
    writeln('  - No traditional "update" - create new structures'),

    halt(0).

% Helper predicates

% Sum list
sum_list([], 0).
sum_list([H|T], Sum) :-
    sum_list(T, RestSum),
    Sum is H + RestSum.

% Map double
map_double([], []).
map_double([H|T], [H2|T2]) :-
    H2 is H * 2,
    map_double(T, T2).

% Filter evens
filter_evens([], []).
filter_evens([H|T], [H|FilteredT]) :-
    0 is H mod 2,
    !,
    filter_evens(T, FilteredT).
filter_evens([_|T], FilteredT) :-
    filter_evens(T, FilteredT).

File: lessons/05-data-structures/data_structures.hs
-- Lesson 5: Data Structures in Haskell
-- Demonstrates immutable lists, tuples, Maps, Sets

import qualified Data.Map as Map
import qualified Data.Set as Set

main :: IO ()
main = do
    putStrLn "=== Haskell Data Structures ===\n"

    -- 1. Lists (immutable, linked)
    putStrLn "1. Lists (Immutable, Linked):"
    let numbers = [1, 2, 3, 4, 5]
    putStrLn $ "  Original: " ++ show numbers

    -- "Modifying" creates new list
    let withZero = 0 : numbers  -- Prepend
    putStrLn $ "  Prepend 0: " ++ show withZero
    putStrLn $ "  Original unchanged: " ++ show numbers

    let withSix = numbers ++ [6]  -- Append (expensive O(n))
    putStrLn $ "  Append 6: " ++ show withSix

    -- List operations
    putStrLn "\n  List operations:"
    putStrLn $ "  head: " ++ show (head numbers)
    putStrLn $ "  tail: " ++ show (tail numbers)
    putStrLn $ "  take 3: " ++ show (take 3 numbers)
    putStrLn $ "  drop 2: " ++ show (drop 2 numbers)

    -- 2. List comprehensions
    putStrLn "\n2. List Comprehensions:"
    let squared = [x * x | x <- numbers]
    putStrLn $ "  Squared: " ++ show squared

    let evens = [x | x <- [1..10], even x]
    putStrLn $ "  Evens 1-10: " ++ show evens

    -- 3. Tuples (immutable, fixed size)
    putStrLn "\n3. Tuples (Immutable, Fixed Size):"
    let point = (3, 4)
    putStrLn $ "  Point: " ++ show point
    putStrLn $ "  fst: " ++ show (fst point)
    putStrLn $ "  snd: " ++ show (snd point)

    let person = ("Alice", 30, "NYC")
    putStrLn $ "  Person: " ++ show person

    -- Pattern matching on tuples
    let (name, age, city) = person
    putStrLn $ "  Destructured: " ++ name ++ ", " ++ show age ++ ", " ++ city

    -- 4. Maps (immutable)
    putStrLn "\n4. Maps (Immutable):"
    let personMap = Map.fromList [("name", "Alice"), ("age", "30"), ("city", "NYC")]
    putStrLn $ "  Person map: " ++ show personMap

    putStrLn $ "  Lookup 'name': " ++ show (Map.lookup "name" personMap)

    -- "Modifying" creates new map
    let updated = Map.insert "age" "31" personMap
    putStrLn $ "  After insert age=31: " ++ show updated
    putStrLn $ "  Original unchanged: " ++ show personMap

    let deleted = Map.delete "city" personMap
    putStrLn $ "  After delete city: " ++ show deleted

    -- 5. Sets (immutable, unique)
    putStrLn "\n5. Sets (Immutable, Unique):"
    let numbersSet = Set.fromList [1, 2, 3, 4, 5]
    putStrLn $ "  Set: " ++ show numbersSet

    let withSix' = Set.insert 6 numbersSet
    let withThree = Set.insert 3 numbersSet  -- Already exists
    putStrLn $ "  After insert 6: " ++ show withSix'
    putStrLn $ "  After insert 3 (duplicate): " ++ show withThree

    putStrLn $ "  Member 3: " ++ show (Set.member 3 numbersSet)
    putStrLn $ "  Member 10: " ++ show (Set.member 10 numbersSet)

    -- Set operations
    let evensSet = Set.fromList [2, 4, 6, 8]
    let oddsSet = Set.fromList [1, 3, 5, 7]
    putStrLn $ "  Union: " ++ show (Set.union evensSet oddsSet)
    putStrLn $ "  Intersection: " ++ show (Set.intersection evensSet oddsSet)
    putStrLn $ "  Difference: " ++ show (Set.difference evensSet oddsSet)

    -- 6. Structural sharing
    putStrLn "\n6. Structural Sharing (Efficiency):"
    putStrLn "  Immutable structures share data internally"
    let list1 = [2, 3, 4, 5]
    let list2 = 1 : list1  -- [1, 2, 3, 4, 5]
    putStrLn $ "  list1: " ++ show list1
    putStrLn $ "  list2 = 1 : list1: " ++ show list2
    putStrLn "  list2 shares nodes with list1 (only 1 new node created)"

    -- 7. Infinite lists (lazy evaluation)
    putStrLn "\n7. Infinite Lists (Lazy Evaluation):"
    let naturals = [1..]
    putStrLn $ "  First 10 naturals: " ++ show (take 10 naturals)

    let fibs = 0 : 1 : zipWith (+) fibs (tail fibs)
    putStrLn $ "  First 10 Fibonacci: " ++ show (take 10 fibs)

    -- 8. Pattern matching on lists
    putStrLn "\n8. Pattern Matching on Lists:"
    putStrLn $ "  describeList []: " ++ describeList []
    putStrLn $ "  describeList [1]: " ++ describeList [1]
    putStrLn $ "  describeList [1,2,3]: " ++ describeList [1,2,3]

    -- 9. Common list functions
    putStrLn "\n9. Common List Functions:"
    putStrLn $ "  length: " ++ show (length numbers)
    putStrLn $ "  reverse: " ++ show (reverse numbers)
    putStrLn $ "  sum: " ++ show (sum numbers)
    putStrLn $ "  product: " ++ show (product numbers)
    putStrLn $ "  maximum: " ++ show (maximum numbers)
    putStrLn $ "  minimum: " ++ show (minimum numbers)

    -- 10. Key insights
    putStrLn "\n10. Key Insights:"
    putStrLn "  - All data structures are IMMUTABLE"
    putStrLn "  - 'Modifications' create new structures"
    putStrLn "  - Structural sharing makes this efficient"
    putStrLn "  - No aliasing bugs possible"
    putStrLn "  - Thread-safe by default"
    putStrLn "  - Lazy evaluation enables infinite structures"

-- Helper function
describeList :: [a] -> String
describeList [] = "Empty"
describeList [_] = "Singleton"
describeList [_, _] = "Pair"
describeList _ = "Many elements"
